File: go/spec/absolute_path_test.go
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package spec

import (
	"fmt"
	"strings"
	"testing"

	"github.com/attic-labs/noms/go/chunks"
	"github.com/attic-labs/noms/go/datas"
	"github.com/attic-labs/noms/go/hash"
	"github.com/attic-labs/noms/go/types"
	"github.com/attic-labs/testify/assert"
)

func TestAbsolutePathToAndFromString(t *testing.T) {
	assert := assert.New(t)

	test := func(str string) {
		p, err := NewAbsolutePath(str)
		assert.NoError(err)
		assert.Equal(str, p.String())
	}

	h := types.Number(42).Hash() // arbitrary hash
	test(fmt.Sprintf("foo.bar[#%s]", h.String()))
	test(fmt.Sprintf("#%s.bar[42]", h.String()))
}

func TestAbsolutePaths(t *testing.T) {
	assert := assert.New(t)

	s0, s1 := types.String("foo"), types.String("bar")
	list := types.NewList(s0, s1)
	emptySet := types.NewSet()

	db := datas.NewDatabase(chunks.NewMemoryStore())
	db.WriteValue(s0)
	db.WriteValue(s1)
	db.WriteValue(list)
	db.WriteValue(emptySet)

	var err error
	ds := db.GetDataset("ds")
	ds, err = db.CommitValue(ds, list)
	assert.NoError(err)
	head := ds.Head()

	resolvesTo := func(exp types.Value, str string) {
		p, err := NewAbsolutePath(str)
		assert.NoError(err)
		act := p.Resolve(db)
		if exp == nil {
			assert.Nil(act)
		} else {
			assert.True(exp.Equals(act), "%s Expected %s Actual %s", str, types.EncodedValue(exp), types.EncodedValue(act))
		}
	}

	resolvesTo(head, "ds")
	resolvesTo(emptySet, "ds.parents")
	resolvesTo(list, "ds.value")
	resolvesTo(s0, "ds.value[0]")
	resolvesTo(s1, "ds.value[1]")
	resolvesTo(head, "#"+head.Hash().String())
	resolvesTo(list, "#"+list.Hash().String())
	resolvesTo(s0, "#"+s0.Hash().String())
	resolvesTo(s1, "#"+s1.Hash().String())
	resolvesTo(s0, "#"+list.Hash().String()+"[0]")
	resolvesTo(s1, "#"+list.Hash().String()+"[1]")

	resolvesTo(nil, "foo")
	resolvesTo(nil, "foo.parents")
	resolvesTo(nil, "foo.value")
	resolvesTo(nil, "foo.value[0]")
	resolvesTo(nil, "#"+types.String("baz").Hash().String())
	resolvesTo(nil, "#"+types.String("baz").Hash().String()+"[0]")
}

func TestReadAbsolutePaths(t *testing.T) {
	assert := assert.New(t)

	s0, s1 := types.String("foo"), types.String("bar")
	list := types.NewList(s0, s1)

	db := datas.NewDatabase(chunks.NewMemoryStore())
	ds := db.GetDataset("ds")
	ds, err := db.CommitValue(ds, list)
	assert.NoError(err)

	vals, err := ReadAbsolutePaths(db, "ds.value[0]", "ds.value[1]")
	assert.NoError(err)

	assert.Equal(2, len(vals))
	assert.Equal("foo", string(vals[0].(types.String)))
	assert.Equal("bar", string(vals[1].(types.String)))

	vals, err = ReadAbsolutePaths(db, "!!#")
	assert.Nil(vals)
	assert.Equal("Invalid input path '!!#'", err.Error())

	vals, err = ReadAbsolutePaths(db, "invalid.monkey")
	assert.Nil(vals)
	assert.Equal("Input path 'invalid.monkey' does not exist in database", err.Error())
}

func TestAbsolutePathParseErrors(t *testing.T) {
	assert := assert.New(t)

	test := func(path, errMsg string) {
		p, err := NewAbsolutePath(path)
		assert.Equal(AbsolutePath{}, p)
		assert.Error(err)
		assert.Equal(errMsg, err.Error())
	}

	test("", "Empty path")
	test(".foo", "Invalid dataset name: .foo")
	test(".foo.bar.baz", "Invalid dataset name: .foo.bar.baz")
	test("#", "Invalid hash: ")
	test("#abc", "Invalid hash: abc")
	invHash := strings.Repeat("z", hash.StringLen)
	test("#"+invHash, "Invalid hash: "+invHash)
}

File: js/noms/src/absolute-path-test.js
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

// @flow

import {assert} from 'chai';
import {suite, test} from 'mocha';
import {equals} from './compare.js';

import {invariant, notNull} from './assert.js';
import AbsolutePath from './absolute-path.js';
import {getHash} from './get-hash.js';
import {stringLength} from './hash.js';
import List from './list.js';
import Set from './set.js';
import {TestDatabase} from './test-util.js';
import type Value from './value.js';

suite('AbsolutePath', () => {
  test('to and from string', () => {
    const t = (str: string) => {
      const p = AbsolutePath.parse(str);
      assert.strictEqual(str, p.toString());
    };

    const h = getHash(42); // arbitrary hash
    t(`foo.bar[#${h.toString()}]`);
    t(`#${h.toString()}.bar[42]`);
  });

  test('absolute paths', async () => {
    const s0 = 'foo', s1 = 'bar';
    const list = new List([s0, s1]);
    const emptySet = new Set();

    const db = new TestDatabase();
    db.writeValue(s0);
    db.writeValue(s1);
    db.writeValue(list);
    db.writeValue(emptySet);

    let ds = db.getDataset('ds');
    ds = await db.commit(ds, list);
    const head = await ds.head();
    invariant(head);

    const resolvesTo = async (exp: Value | null, str: string) => {
      const p = AbsolutePath.parse(str);
      const act = await notNull(p).resolve(db);
      if (exp === null) {
        assert.strictEqual(null, act);
      } else if (act === null) {
        assert.isTrue(false, `Failed to resolve ${str}`);
      } else {
        assert.isTrue(equals(exp, act));
      }
    };

    await resolvesTo(head, 'ds');
    await resolvesTo(emptySet, 'ds.parents');
    await resolvesTo(list, 'ds.value');
    await resolvesTo(s0, 'ds.value[0]');
    await resolvesTo(s1, 'ds.value[1]');
    await resolvesTo(head, '#' + getHash(head).toString());
    await resolvesTo(list, '#' + getHash(list).toString());
    await resolvesTo(s0, `#${getHash(s0).toString()}`);
    await resolvesTo(s1, `#${getHash(s1).toString()}`);
    await resolvesTo(s0, `#${getHash(list).toString()}[0]`);
    await resolvesTo(s1, `#${getHash(list).toString()}[1]`);

    await resolvesTo(null, 'foo');
    await resolvesTo(null, 'foo.parents');
    await resolvesTo(null, 'foo.value');
    await resolvesTo(null, 'foo.value[0]');
    await resolvesTo(null, `#${getHash('baz').toString()}`);
    await resolvesTo(null, `#${getHash('baz').toString()}[0]`);
  });

  test('parse errors', () => {
    const t = (path: string, exp: string) => {
      let act = '';
      try {
        AbsolutePath.parse(path);
      } catch (e) {
        assert.instanceOf(e, SyntaxError);
        act = e.message;
      }
      assert.strictEqual(exp, act);
    };

    t('', 'Empty path');
    t('.foo', 'Invalid dataset name: .foo');
    t('.foo.bar.baz', 'Invalid dataset name: .foo.bar.baz');
    t('#', 'Invalid hash: ');
    t('#abc', 'Invalid hash: abc');
    const invalidHash = new Array(stringLength).join('z');
    t(`#${invalidHash}`, `Invalid hash: ${invalidHash}`);
  });
});
